pkg/services/live/runstream/manager.go

```diff
@@ -0,0 +1,289 @@
+	return func(sm *Manager) {
+		sm.checkInterval = interval
+		sm.maxChecks = maxChecks
+	}
+}
+
+const (
+	defaultCheckInterval = 5 * time.Second
+	defaultMaxChecks     = 3
+)
+
+// NewManager creates new Manager.
+func NewManager(packetSender StreamPacketSender, presenceGetter PresenceGetter, opts ...ManagerOption) *Manager {
+	sm := &Manager{
+		streams:        make(map[string]chan struct{}),
+		packetSender:   packetSender,
+		presenceGetter: presenceGetter,
+		registerCh:     make(chan submitRequest),
+		closedCh:       make(chan struct{}),
+		checkInterval:  defaultCheckInterval,
+		maxChecks:      defaultMaxChecks,
+	}
+	for _, opt := range opts {
+		opt(sm)
+	}
+	return sm
+}
+
+func (s *Manager) stopStream(sr streamRequest, cancelFn func()) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	closeCh, ok := s.streams[sr.Channel]
+	if !ok {
+		return
+	}
+	delete(s.streams, sr.Channel)
+	cancelFn()
+	close(closeCh)
+}
+
+func (s *Manager) watchStream(ctx context.Context, cancelFn func(), sr streamRequest) {
+	numNoSubscribersChecks := 0
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-time.After(s.checkInterval):
+			numSubscribers, err := s.presenceGetter.GetNumSubscribers(sr.Channel)
+			if err != nil {
+				logger.Error("Error checking num subscribers", "channel", sr.Channel, "path", sr.Path)
+				continue
+			}
+			if numSubscribers > 0 {
+				// reset counter since channel has active subscribers.
+				numNoSubscribersChecks = 0
+				continue
+			}
+			numNoSubscribersChecks++
+			if numNoSubscribersChecks >= s.maxChecks {
+				logger.Debug("Stop stream since no active subscribers", "channel", sr.Channel, "path", sr.Path)
+				s.stopStream(sr, cancelFn)
+				return
+			}
+		}
+	}
+}
+
+const streamDurationThreshold = 100 * time.Millisecond
+const coolDownDelay = 100 * time.Millisecond
+const maxDelay = 5 * time.Second
+
+func getDelay(numErrors int) time.Duration {
+	if numErrors == 0 {
+		return 0
+	}
+	delay := coolDownDelay * time.Duration(math.Pow(2, float64(numErrors)))
+	if delay > maxDelay {
+		return maxDelay
+	}
+	return delay
+}
+
+// run stream until context canceled or stream finished without an error.
+func (s *Manager) runStream(ctx context.Context, cancelFn func(), sr streamRequest) {
+	var numFastErrors int
+	var delay time.Duration
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		default:
+		}
+		startTime := time.Now()
+		err := sr.StreamRunner.RunStream(
+			ctx,
+			&backend.RunStreamRequest{
+				PluginContext: sr.PluginContext,
+				Path:          sr.Path,
+			},
+			newStreamSender(sr.Channel, s.packetSender),
+		)
+		if err != nil {
+			if errors.Is(ctx.Err(), context.Canceled) {
+				logger.Debug("Stream cleanly finished", "path", sr.Path)
+				return
+			}
+			// Best effort to cool down re-establishment process. We don't have a
+			// nice way to understand whether we really need to wait here - so relying
+			// on RunStream duration time.
+			if time.Since(startTime) < streamDurationThreshold {
```

Comment:
this is where introduced backoff starts - sorry for such a pr, I better move code between packages in separate pr...